src/test/java/org/example/bookshop/controller/CategoryControllerTest.java

```diff
@@ -0,0 +1,216 @@
+                .andExpect(status().isOk())
+                .andReturn();
+        // Then
+        BookDtoWithoutCategoryIds[] actual = objectMapper.readValue(
+                result.getResponse().getContentAsByteArray(), BookDtoWithoutCategoryIds[].class);
+        List<BookDtoWithoutCategoryIds> actualList = Arrays.stream(actual).toList();
+        Assertions.assertEquals(expected.size(), actualList.size());
+        Assertions.assertTrue(IntStream.range(0, expected.size())
+                .allMatch(i -> EqualsBuilder.reflectionEquals(expected.get(i), actualList.get(i))));
+    }
+
+    @Test
+    @WithMockUser(roles = {"ADMIN"})
+    @DisplayName("Update category by valid id")
+    void updateCategoryById_ValidRequestDtoAndId_ReturnCategoryDto() throws Exception {
+        // Given
+        CategoryDto updatedCategoryDto = new CategoryDto();
+        updatedCategoryDto.setName(CATEGORY_NAME + NEW_CATEGORY_PARAM);
+        updatedCategoryDto.setDescription(CATEGORY_DESCRIPTION + NEW_CATEGORY_PARAM);
+        String jsonRequest = objectMapper.writeValueAsString(updatedCategoryDto);
+        // When
+        MvcResult result = mockMvc.perform(
+                        put("/categories/{id}", FIRST_ID)
+                                .content(jsonRequest)
+                                .contentType(MediaType.APPLICATION_JSON)
+                )
+                .andExpect(status().isOk())
+                .andReturn();
+        // Then
+        CategoryDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), CategoryDto.class);
+        EqualsBuilder.reflectionEquals(updatedCategoryDto, actual);
+    }
+
+    @Test
+    @WithMockUser(roles = {"ADMIN"})
+    @DisplayName("Delete category by valid id")
+    void deleteCategoryById_ValidId_CallDeleteMethodOnce() throws Exception {
+        // When
+        mockMvc.perform(
+                        MockMvcRequestBuilders.delete("/categories/{id}", FIRST_ID))
+                .andExpect(status().isNoContent());
+    }
+
+    private List<BookDtoWithoutCategoryIds> getCategoryIds() {
+        BookDtoWithoutCategoryIds bookDto = new BookDtoWithoutCategoryIds();
+        bookDto.setId(SECOND_ID);
+        bookDto.setTitle(BOOK_TITLE + UNIQUE_PARAM);
+        bookDto.setAuthor(BOOK_AUTHOR + UNIQUE_PARAM);
+        bookDto.setIsbn(BOOK_ISBN + UNIQUE_PARAM);
+        bookDto.setPrice(BOOK_PRICE);
+
+        BookDtoWithoutCategoryIds secondBookDto = new BookDtoWithoutCategoryIds();
+        secondBookDto.setId(THIRD_ID);
+        secondBookDto.setTitle(UNIQUE_PARAM + BOOK_TITLE);
+        secondBookDto.setAuthor(UNIQUE_PARAM + BOOK_AUTHOR);
+        secondBookDto.setIsbn(UNIQUE_PARAM + BOOK_ISBN);
+        secondBookDto.setPrice(BOOK_PRICE);
+
+        return List.of(bookDto, secondBookDto);
+    }
+
+    private List<CategoryDto> createCategoriesDto() {
+        CategoryDto secondCategoryDto = new CategoryDto();
```

Comment:
It's better to create TestUtil class, create static methods to generate test data and provide constants.